### PR TITLE
DateTime: better integration with chrono, support year 0001+

### LINKED ISSUE #1132: Add TimeSpan in Core SDK
```c  
Azure::Storage::Blobs::BreakBlobContainerLeaseOptions.BreakPeriod
```
#
Option 1: Add duration type.
Option 2: Add `InSeconds` suffix.

### LINKED ISSUE #1174: DateTime: Change epoch to be year 1, allow printing and parsing from year 1 to 1601, in addition to 1601 to 9999

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #1174.
Closes #1132.

### CODE REVIEW

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (Jinming-Hu) [inline sdk/core/azure-core/inc/azure/core/datetime.hpp:106]: explicit

**Comment by Jinming-Hu** [comment]
@antkmsft Because you made changes to storage code in this PR, we'd appreciate if you can wait for storage team's approval before merging it next time.

**Comment by antkmsft** [comment]
@Jinming-Hu , sorry, I'll wait next time. It is basically a syntactic update - DateTime no longer has Now() method.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:106]: We discussed it with Jeffrey and decided to make it implicit. Reasons:
1. `DateTime dt = std::chrono::system_clock::now()` - convenient for customers
2. No harm in converting from system_clock - the system_clock's datetime is, realistically, always representable in terms of date range (between year 0001 and 9999; on POSIX system_clock has 1ns precision, while DateTime's is 100ns, but benefits (item 1) seem to outweigh costs.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:106]: Sure. I brought this up only because we ever got a conclusion in guidelines meeting that constructors should be explicit unless specifically permitted.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:106]: Thank you, good point, let me to bring this up during our next meeting.
I believe I have justification, since this whole effort with deriving from std::chrono::time_point<Clock> is so that our DateTime mimics chrono as close as possible, so this class stands in the intersection of standard library and Azure SDK integrating them and providing syntactic convenience for our users. Let's see what people say, I hope they agree, but it will be good to get formal approval and confirmation from broader audience, thank you for bringing this up.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:106]: I also agree it's better to be implicit than explicit for this one, and I see Jeff signed off. So I think it's okay.

**Comment by antkmsft** [comment]
@Jinming-Hu, just checking, did you relied on default-constructed `DateTime` being `1601-01-01`? It is `0001-01-01` now.
Formally, for the previous version, we deliberately never documented what the default-constructed DateTime's date might be, however it is hard to keep that a secret. We do document now, BTW, we added in doxygen comment in this PR.

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu, just checking, did you relied on default-constructed `DateTime` being `1601-01-01`? It is `0001-01-01` now.
> Formally, for the previous version, we deliberately never documented what the default-constructed DateTime's date might be, however it is hard to keep that a secret. We do document now, BTW, we added in doxygen comment in this PR.

No, we don't reply on that.

Actually I tend to make a default-constructed `DateTime` to be undefined value, or uninitialized state. This aligns with `utility::datetime` impl in cpprest sdk.

**Comment by Jinming-Hu** [comment]
Well I feel really bad after I saw the explicit conversion to `std:chrono::time_point`, when I was trying to figure out a bug in comparison operators. So question: when is the explicit conversion function is used?

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/datetime.hpp:51]: Looks like this PR/change is causing vcpkg CI build to fail:
https://github.com/microsoft/vcpkg/pull/15635

@antkmsft If we can't fix it easily (and if it requires more investigation), should we revert this change fully and try a different approach?

File: sdk/core/azure-core/test/ut/datetime.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/datetime.cpp:509]: Why did we comment out these test cases?

File: sdk/core/azure-core/src/datetime.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/datetime.cpp:115]: nit:
```suggestion
constexpr int8_t WeekDayAndMonthDayOfYear(int8_t* month, int8_t* day, int16_t year, int16_t dayOfYear)
```

File: sdk/core/azure-core/src/datetime.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/datetime.cpp:120]: This logic is difficult to parse and understand. Add a comment explaining what you are trying to do here.

File: sdk/core/azure-core/src/datetime.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/datetime.cpp:236]: Why is this `static_cast` safe to do and not lossy?

File: sdk/core/azure-core/src/datetime.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/datetime.cpp:364]: I don't quite understand what the purpose of this is. What does `GetSystemClockEpoch` do and why is it needed?

**Review by ahsonkhan — COMMENTED** [review]
Partial review. Will continue reviewing the rest of the implementation later.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:34]: I'll sneak this in to one of my other PRs.

File: sdk/core/azure-core/test/ut/datetime.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/datetime.cpp:509]: Because it is valid offset, per spec, and on .NET

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:50]: It is okay to inherit or typedef an internal type. It is good that the clock is not cluttering the public namespace - it is useless for the customer. Anything that is available via DateTime class is open to customer, including via inherited type, but it is not ok to go and consume the type via `::Details`. We talked about that. And I think you agreed
If you disagree, please explain, what benefit for the customer would be to have the clock as public class (not via DateTime::clock, but as Azure::Core::Clock)

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:71]: It is private detali

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:93]: it is private detail

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:100]: It is `DateTime::time_point`, which is typedeffed to `std::chrono::time_point<Azure::Core::Details::Clock>`.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:116]: year 0000 or 10000.

File: sdk/core/azure-core/inc/azure/core/datetime.hpp
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/datetime.hpp:196]: it is not a declaration, but implementation. It does not place the declared name outside of the Details namespace.

File: sdk/core/azure-core/src/datetime.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/datetime.cpp:52]: I don't think it is complicated, and I think it has value.

File: sdk/core/azure-core/src/datetime.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/datetime.cpp:236]: Purely from usage / points of invocation we know that. This is a helper function that is carefully used within this class.
If there are problems, unit tests will find, by testing broader behavior.

File: sdk/core/azure-core/src/datetime.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/datetime.cpp:364]: It is initialization for static class member.
System clock's zero could be 1970, or 1601. Standard does not specify. Which is why GetSystemClockEpoch() finds it out, and returns it in terms of Azure::Core::DateTime, where it is certainly not zero.
Static private class member SystemClockEpoch is needed so that conversion from `std::chrono::system_clock::time_point` works, and is inlineable:
```
DateTime(std::chrono::system_clock::time_point const& systemTime)
        : DateTime(
            SystemClockEpoch + std::chrono::duration_cast<duration>(systemTime.time_since_epoch()))
    {
    }
```

**Pull request merged**
